lib/ui/painting.dart

```diff
@@ -3850,8 +3853,20 @@ class FragmentShader extends Shader {
         'FragmentShader floatUniforms size: ${floatUniforms.length} must match given shader uniform count: $_uniformFloatCount.');
     }
     _update(floatUniforms);
+    _hashCode = hashValues(_spirvHash, hashList(floatUniforms));
   }
 
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    return other is FragmentShader && other._hashCode == _hashCode;
```

Comment:
This is statistically pretty strong, but theoretically incorrect. How often does someone recreate one of these from scratch with the exact same values?

The correct implementation would be to save the 2 value buffers and compare them directly rather than comparing their hashes. If that is too expensive, or costs too much memory, then this may be a case where we rely on `identical` behaviors for hash/equals in the absence of better information.

Reply:
I don't think it is too expensive, but I'd be in favor of dropping this PR and relying on `identical` behavior if that's acceptable for the engine. The 2 value buffers are immediately sent to the engine and would serve no other purpose than for equality-testing, and the recommended approach here should be to make use of the same FragmentShader object if you need the same shader in multiple places. That said, if it makes sense to keep them around I can update this to follow your suggestions.

Reply:
"The engine" has no requirements. What we are discussing here is how optimal the implementation is.

For example, in the ShaderMaskLayer we have https://github.com/flutter/flutter/blob/870f5e88d298b001ee4b2e5778ec975e5b272c2a/packages/flutter/lib/src/rendering/layer.dart#L1853 where the shader is compared against the previously used value to see if that part of the tree should be considered "dirty" and re-added to the scene. If an identity test works fine in practice for that test, then this particular line of code will be doing a reasonably optimal job. If applications are written to construct the FragmentShader objects on the fly, then it will not work well.

Another area where this comes up is in recording Picture objects where one of the rendering operations used one of these shaders. If the render object for a widget is being repainted due to another widget changing state and the two are not separated by a RepaintBoundary, then the Picture will be reconstructed. If the `Paint.shader = ...` call reconstructs the FragmentShader on the fly then it will be a new object and when the engine determines which parts of the scene need to be repainted and which can be cached then it can only use an identity == comparison on the SkShader object to determine the stability of that picture. (This type of work is coming in the near future, but isn't yet a part of our optimizations).

The latter consideration would be moot if SkShader implemented ==, but it doesn't.

Deciding how much the ==/hashcode contract of the FragmentShader might impact the optimal behavior in those cases depends a lot on how the application is written. If it were quick and frugal to be able to do a test to verify that the two non-identity objects are equivalent then that would be the safest solution, but if it requires more retained memory then we have to balance that out against the practical likelihood of these situations.

Reply:
Thanks for the code pointers - I understand - what I meant by 'acceptable' was specifically in regards to the new optimizations you are working on that you mentioned in our other thread. I'm going to resolve this thread and continue this down below - I agree with your latest suggestions.